aggregator/aggregator.go

```diff
@@ -732,16 +743,31 @@ func (a *Aggregator) tryGenerateBatchProof(ctx context.Context, prover *prover.P
 	return true, nil
 }
 
-// verifyProofTimeReached returns if we have reached the timeout to verify a
+// canVerifyProof returns if we have reached the timeout to verify a
 // proof.
-func (a *Aggregator) verifyProofTimeReached() bool {
-	a.TimeSendFinalProofMutex.RLock()
-	defer a.TimeSendFinalProofMutex.RUnlock()
-	return a.TimeSendFinalProof.Before(time.Now())
+func (a *Aggregator) canVerifyProof() bool {
+	a.TimeSendFinalProofMutex.Lock()
+	defer a.TimeSendFinalProofMutex.Unlock()
+	if a.TimeSendFinalProof.Before(time.Now()) {
+		if a.verifyingProof {
+			return false
+		}
+		a.verifyingProof = true
+		return true
+	}
+	return false
+}
+
+func (a *Aggregator) enableProofVerification() {
+	a.TimeSendFinalProofMutex.Lock()
+	defer a.TimeSendFinalProofMutex.Unlock()
+	a.verifyingProof = false
+
 }
 
 // resetVerifyProofTime updates the timeout to verify a proof.
 func (a *Aggregator) resetVerifyProofTime() {
+	a.enableProofVerification()
```

Comment:
@agnusmor does this look better for you? a9a81e1b2ff4cca0443b1116364b8bf077a258c1
I don't see a major problem by calling `enableProofVerification` inside `resetVerifyProofTime` as the mutex is unlocked and then locked again on the next instruction. Also `resetVerifyProofTime` can't be called in parallel.

Anyway I saw a small drawback in  this implementation. Now in `canVerifyProof` we lock the `TimeSendFinalProof` variable for all the readers. Considering that most of the time the provers will only check that variable and they will not enter the `if` statement (because the timeout is not reached yet), this would unnecessarily slow down the provers. Preformance is not a key point now, but I still wanted to make this clear (that's why I was thinking about having a separate mutex).

Reply:
For me it's ok as a quick fix now. We can review it later. I understand what you said, but really this checks inside the lock are very fast and this will not have any impact in the performance. Anyway we can review this later with more time.